--- network-3.1.2.5-r0/network.cabal	2024-11-14 15:47:11.934127415 +1300
+++ network-3.1.2.5/network.cabal	2024-11-14 15:47:11.943408736 +1300
@@ -1,6 +1,7 @@
 cabal-version:  1.18
 name:           network
 version:        3.1.2.5
+x-revision: 1
 license:        BSD3
 license-file:   LICENSE
 maintainer:     Kazu Yamamoto, Evan Borden
@@ -87,7 +88,7 @@
     Network.Socket.Unix
 
   build-depends:
-    base >= 4.9 && < 5,
+    base >= 4.9 && < 4.17,
     bytestring >= 0.10 && < 0.12,
     deepseq,
     directory
@@ -146,7 +147,7 @@
   --     both packages to a small enough version range.
   build-tools: hspec-discover >= 2.6
   build-depends:
-    base >= 4.9 && < 5,
+    base >= 4.9 && < 4.17,
     bytestring,
     directory,
     HUnit,
@@ -163,7 +164,7 @@
   type: exitcode-stdio-1.0
 
   build-depends:
-    base >= 4.9 && < 5,
+    base >= 4.9 && < 4.17,
     doctest >= 0.10.1,
     network
 
