diff -X ../utils/updatePatches.excludes -Nru upstream/System/Posix/Files/Common.hsc boot/System/Posix/Files/Common.hsc
--- upstream/System/Posix/Files/Common.hsc	2001-09-09 03:46:40.000000000 +0200
+++ boot/System/Posix/Files/Common.hsc	2021-06-10 22:32:23.930000000 +0200
@@ -520,7 +520,7 @@
 -- Note: calls @ftruncate@.
 setFdSize :: Fd -> FileOffset -> IO ()
 setFdSize (Fd fd) off =
-  throwErrnoIfMinus1_ "setFdSize" (c_ftruncate fd off)
+  throwErrnoIfMinus1_ "setFdSize" (c_ftruncate fd (fromIntegral off))
 
 -- -----------------------------------------------------------------------------
 -- pathconf()/fpathconf() support
diff -X ../utils/updatePatches.excludes -Nru upstream/System/Posix/IO/Common.hsc boot/System/Posix/IO/Common.hsc
--- upstream/System/Posix/IO/Common.hsc	2001-09-09 03:46:40.000000000 +0200
+++ boot/System/Posix/IO/Common.hsc	2021-06-10 22:32:23.930000000 +0200
@@ -281,7 +281,7 @@
 -- | May throw an exception if this is an invalid descriptor.
 fdSeek :: Fd -> SeekMode -> FileOffset -> IO FileOffset
 fdSeek (Fd fd) mode off =
-  throwErrnoIfMinus1 "fdSeek" (Base.c_lseek fd off (mode2Int mode))
+  throwErrnoIfMinus1 "fdSeek" (fmap fromIntegral $ Base.c_lseek fd (fromIntegral off) (mode2Int mode))
 
 -- -----------------------------------------------------------------------------
 -- Locking
diff -X ../utils/updatePatches.excludes -Nru upstream/System/Posix/Process/Internals.hsc boot/System/Posix/Process/Internals.hsc
--- upstream/System/Posix/Process/Internals.hsc	2001-09-09 03:46:40.000000000 +0200
+++ boot/System/Posix/Process/Internals.hsc	2021-06-10 22:32:23.930000000 +0200
@@ -1,5 +1,6 @@
 {-# LANGUAGE CApiFFI #-}
 {-# LANGUAGE Trustworthy #-}
+{-# LANGUAGE CPP #-}
 
 module System.Posix.Process.Internals (
        pPrPr_disableITimers, c_execvpe,
@@ -9,7 +10,12 @@
 import Foreign.C
 import System.Exit
 import System.IO.Error
+
+#ifdef ghcjs_HOST_OS
+type Signal = CInt
+#else
 import GHC.Conc (Signal)
+#endif
 
 -- | The exit status of a process
 data ProcessStatus
diff -X ../utils/updatePatches.excludes -Nru upstream/config.sub boot/config.sub
--- upstream/config.sub	2001-09-09 03:46:40.000000000 +0200
+++ boot/config.sub	2021-06-11 14:49:50.712651100 +0200
@@ -1169,6 +1169,7 @@
 			| arm | arm[lb]e | arme[lb] | armv* \
 			| avr | avr32 \
 			| asmjs \
+			| js \
 			| ba \
 			| be32 | be64 \
 			| bfin | bpf | bs2000 \
@@ -1728,7 +1729,7 @@
 	     | powermax* | dnix* | nx6 | nx7 | sei* | dragonfly* \
 	     | skyos* | haiku* | rdos* | toppers* | drops* | es* \
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
-	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
+	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* | ghcjs* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx*)
 		;;
 	# This one is extra strict with allowed versions
