--- clock-0.8.2-r0/clock.cabal	2024-11-14 15:46:53.708573858 +1300
+++ clock-0.8.2/clock.cabal	2024-11-14 15:46:53.717652971 +1300
@@ -1,6 +1,7 @@
 cabal-version: >= 1.10
 name:          clock
 version:       0.8.2
+x-revision: 1
 stability:     stable
 synopsis:      High-resolution clock functions: monotonic, realtime, cputime.
 description:   A package for convenient access to high-resolution clock and
@@ -68,8 +69,9 @@
 library
     default-language: Haskell2010
     if impl (ghc < 7.6)
-      build-depends:       base >= 4.4 && <= 5, ghc-prim
-    build-depends:       base >= 2 && <= 5
+      build-depends:     base >= 4.4 && <= 5, ghc-prim
+    build-depends:       base >= 4.4 && <= 5
+                           -- clock uses LANGUAGE DeriveGeneric, available only from GHC 7.2
     exposed-modules:     System.Clock
     default-extensions:          DeriveGeneric
                          DeriveDataTypeable
