--- unix-compat-0.5.3-r0/unix-compat.cabal	2024-11-14 15:47:20.306719749 +1300
+++ unix-compat-0.5.3/unix-compat.cabal	2024-11-14 15:47:20.315982445 +1300
@@ -1,5 +1,6 @@
 name:           unix-compat
 version:        0.5.3
+x-revision: 3
 synopsis:       Portable POSIX-compatibility layer.
 description:    This package provides portable implementations of parts
                 of the unix package. This package re-exports the unix
@@ -45,7 +46,7 @@
       cbits/mktemp.c
 
     extra-libraries: msvcrt
-    build-depends: Win32 >= 2.5.0.0
+    build-depends: Win32 >= 2.5.0.0 && < 2.12
 
     if flag(old-time)
       build-depends: old-time >= 1.0.0.0 && < 1.2.0.0
@@ -64,7 +65,7 @@
       System.PosixCompat.Internal.Time
 
   else
-    build-depends: unix >= 2.4 && < 2.9
+    build-depends: unix >= 2.6 && < 2.8
     include-dirs: include
     includes: HsUnixCompat.h
     install-includes: HsUnixCompat.h
